File: src/main/java/com/boxfox/core/router/FriendRouter.java
package com.boxfox.core.router;

import com.boxfox.vertx.vertx.router.*;
import com.boxfox.vertx.vertx.service.*;
import com.boxfox.cross.service.FriendService;
import com.google.gson.Gson;
import io.vertx.core.http.HttpMethod;
import io.vertx.ext.web.RoutingContext;


public class FriendRouter extends AbstractRouter {

    @Service
    private FriendService friendService;
    private Gson gson;

    public FriendRouter() {
        this.gson = new Gson();
    }

    @RouteRegistration(uri = "/search/account/list", method = HttpMethod.GET, auth = true)
    public void search(RoutingContext ctx, @Param String text) {
        friendService.serachUsers(text, res -> {
            ctx.response().end(gson.toJson(res.result()));
        });
    }

    @RouteRegistration(uri = "/search/account", method = HttpMethod.GET, auth = true)
    public void searchAccount(RoutingContext ctx, @Param String uid) {
        friendService.getUser(uid, res -> {
            ctx.response().end(gson.toJson(res.result()));
        });
    }

    @RouteRegistration(uri = "/friend", method = HttpMethod.GET, auth = true)
    public void loadFriends(RoutingContext ctx) {
        String uid = (String) ctx.data().get("uid");
        friendService.loadFriends(uid, res -> {
            ctx.response().end(gson.toJson(res.result()));
        });
    }


    @RouteRegistration(uri = "/friend", method = HttpMethod.PUT, auth = true)
    public void addFriend(RoutingContext ctx, @Param String targetUid) {
        String ownUid = (String) ctx.data().get("uid");
        friendService.addFriend(ownUid, targetUid, res -> {
            if (res.succeeded()) {
                ctx.response().setStatusCode(200).end();
            } else {
                ctx.fail(res.cause());
            }
        });
    }

    @RouteRegistration(uri = "/friend", method = HttpMethod.DELETE, auth = true)
    public void deleteFriend(RoutingContext ctx, @Param String targetUid) {
        String ownUid = (String) ctx.data().get("uid");
        friendService.deleteFriend(ownUid, targetUid, res -> {
            if (res.succeeded()) {
                ctx.response().setStatusCode(200).end();
            } else {
                ctx.fail(res.cause());
            }
        });
    }


}

File: src/main/java/com/boxfox/core/router/FriendRouter.kt
package com.boxfox.core.router

import com.boxfox.cross.service.friend.FriendService
import com.boxfox.vertx.router.AbstractRouter
import com.boxfox.vertx.router.Param
import com.boxfox.vertx.router.RouteRegistration
import com.boxfox.vertx.service.Service
import io.vertx.core.http.HttpMethod
import io.vertx.ext.web.RoutingContext


class FriendRouter : AbstractRouter() {

    @Service
    private lateinit var friendService: FriendService

    @RouteRegistration(uri = "/search/account/list", method = arrayOf(HttpMethod.GET), auth = true)
    fun search(ctx: RoutingContext, @Param(name = "text") text: String) {
        friendService.serachUsers(text).subscribe({ list ->
            ctx.response().end(gson.toJson(list))
        }, {
            ctx.fail(it)
        })
    }

    @RouteRegistration(uri = "/search/account", method = arrayOf(HttpMethod.GET), auth = true)
    fun searchAccount(ctx: RoutingContext, @Param(name = "uid") uid: String) {
        friendService.getUser(uid).subscribe({ user ->
            ctx.response().end(gson.toJson(user))
        }, {
            ctx.fail(it)
        })
    }

    @RouteRegistration(uri = "/friend", method = arrayOf(HttpMethod.GET), auth = true)
    fun loadFriends(ctx: RoutingContext) {
        val uid = ctx.data()["uid"] as String
        friendService.loadFriends(uid).subscribe({ list ->
            ctx.response().end(gson.toJson(list))
        }, {
            ctx.fail(it)
        })
    }


    @RouteRegistration(uri = "/friend", method = arrayOf(HttpMethod.PUT), auth = true)
    fun addFriend(ctx: RoutingContext, @Param(name = "uid") targetUid: String) {
        val ownUid = ctx.data()["uid"] as String
        friendService.addFriend(ownUid, targetUid).subscribe({
            ctx.response().setStatusCode(200).end()
        }, {
            ctx.fail(it)
        })
    }

    @RouteRegistration(uri = "/friend", method = arrayOf(HttpMethod.DELETE), auth = true)
    fun deleteFriend(ctx: RoutingContext, @Param(name = "uid") targetUid: String) {
        val ownUid = ctx.data()["uid"] as String
        friendService.deleteFriend(ownUid, targetUid).subscribe({
            ctx.response().setStatusCode(200).end()
        }, {
            ctx.fail(it)
        })
    }


}
